feat: Add support for TypeScript config files (#117)
eslint/rfcs: designs/2024-support-ts-config-files/README.md

- Repo: eslint/eslint
- Start Date: 2024-03-09

> - RFC PR: (leave this empty, to be filled in later)

Reviewer: ```suggestion
- RFC PR: <https://github.com/eslint/rfcs/pull/117>
```

- Authors: [Arya Emami](https://github.com/aryaemami59)

# Add Support for TypeScript Config Files

## Summary

Add support for TypeScript config files (`eslint.config.ts`, `eslint.config.mts`, `eslint.config.cts`)

<!-- One-paragraph explanation of the feature. -->

## Motivation

<!-- Why are we doing this? What use cases does it support? What is the expected
outcome? -->

The primary motivation for adding support for TypeScript configuration files to ESLint is to enhance the developer experience and accommodate the evolving JavaScript ecosystem. As TypeScript's popularity continues to grow, more projects are adopting TypeScript not only for their source code but also for their configuration files. This shift is driven by TypeScript's ability to provide compile-time type checks and IntelliSense. By supporting `eslint.config.ts`, `eslint.config.mts`, and `eslint.config.cts`, ESLint will offer first-class support to TypeScript users, allowing them to leverage these benefits directly within their ESLint configuration.

> ## Detailed Design

Reviewer: How does the feature interact with the [CLI option](https://eslint.org/docs/latest/use/command-line-interface#options) `--config`?

Author: I have tested it, so far it seems to work pretty well actually, especially with v9. I'm probably going to write a bunch of tests as well to see if there are any edge cases but so far so good.

Reviewer: It would be good to explain what the behavior is when specifying TS or non-TS config files of varying file extensions through that option.

Author: It doesn't behave any differently, same as before. You can do `eslint . --config=eslint.config.ts` or `eslint . -c eslint.config.ts` and they just work. Same as with a `eslint.config.js` file.

Reviewer 2: Can you add that into the RFC?

Author: I added it to the open questions, is that fine?

> ## Detailed Design

Reviewer: It would be worth providing an example of a basic TypeScript config file, including the type annotation of the exported object, to illustrate the nice syntax and type-checking it enables.

Author: That's a great idea. I will do that, thanks!

- There should not be extra overhead for JavaScript users. This means this change should not have a significant impact (if any at all) affecting users who use plain JavaScript config files.
- The external tools that are used to parse the config files written in TypeScript should not create side effects. Specifically, it is imperative that these tools do not interfere with Node.js's native module resolution system by hooking into or altering the standard `import/require` mechanisms. This means tools like [`ts-node`](https://github.com/TypeStrong/ts-node) and [`tsx`](https://github.com/privatenumber/tsx) might not be suitable for this purpose.

> So far the tool that seems to be the most suitable for this purpose is [`jiti`](https://github.com/unjs/jiti). It does not introduce side effects and performs well, demonstrating its reliability. It also seems to be more battle-tested given some established frameworks such as [Nuxt](https://github.com/nuxt/nuxt) and [Tailwind CSS](https://github.com/tailwindlabs/tailwindcss) have been using it to load their configuration files.

Reviewer: Could be worth linking to https://www.npmjs.com/package/jiti as well, as this further illustrates its popularity with 10M+ weekly downloads.

Author: It is linked inside the `README.md` do you want me to link it in the description of the issue as well?

Reviewer: I don't see a link to the package on npm? The main point is just to mention its popularity which is important to be aware of when adding a dependency.

Author: Sure thing, I can add the NPM link.

## Drawbacks

<!--
    Why should we *not* do this? Consider why adding this into ESLint
    might not benefit the project or the community. Attempt to think
    about any opposing viewpoints that reviewers might bring up.

    Any change has potential downsides, including increased maintenance
    burden, incompatibility with other tools, breaking existing user
    experience, etc. Try to identify as many potential problems with
    implementing this RFC as possible.
-->

This change will most likely require at least one external tool as either a dependency or a peer dependency.

## Backwards Compatibility Analysis

<!--
    How does this change affect existing ESLint users? Will any behavior
    change for them? If so, how are you going to minimize the disruption
    to existing users?
-->

This goal is to minimize the disruption to existing users. The primary focus is to ensure that the existing JavaScript configuration files continue to work as expected. The changes will only affect TypeScript users who are using TypeScript configuration files. The changes will not affect the existing JavaScript configuration files.

> ## Alternatives

Reviewer: It would be worth listing the several alternatives you considered in your [PR](https://github.com/eslint/eslint/pull/18134).

Author: I do mention them in the `README.md` just not as "alternatives" since the maintainers were very clear that they wanted to steer clear of `import/require` related side effects and both [`tsx`](https://github.com/privatenumber/tsx) and [`ts-node`](https://github.com/TypeStrong/ts-node) take over module imporing and cause issues. In this case they don't seem to be suitable because we are going to manually work with `require.cache` and sometimes they can cause issues with the cache to get out of sync.

Reviewer: I would find it helpful to see these mentioned in the alternatives section. In the alternatives section, you can explicitly mention pros/cons and why you decided against each alternative. This makes it more clear that you considered different options, as opposed to leaving this section blank.

Author: Sure I can do that.

<!--
    What other designs did you consider? Why did you decide against those?

    This section should also include prior art, such as whether similar
    projects have already implemented a similar feature.
-->

## Open Questions

<!--
    This section is optional, but is suggested for a first draft.

    What parts of this proposal are you unclear about? What do you
    need to know before you can finalize this RFC?

    List the questions that you'd like reviewers to focus on. When
    you've received the answers and updated the design to reflect them,
    you can remove this section.
-->

1. How is caching going to work with TypeScript config files?
2. Should we look at the nearest `tsconfig.json` file to determine the module resolution for `eslint.config.ts` files? Most likely not, but it's worth considering.
3. Should we allow some sort of interoperability between JavaScript and TypeScript configuration files? For example, should we allow a TypeScript configuration file to extend a JavaScript configuration file and vice versa?

> 4. Should we allow `eslint.config.ts` to be able to use `export default` as well as `module.exports` (might be related to [TypeScript's automatic Module Detection](https://www.typescriptlang.org/tsconfig#moduleDetection))?

Author: The main concern for me is since we're going to handle transpiling the TS config files, should we make it so that the user can use either `module.exports` or `export default` and have it just work? For reference in the linked PR I set the `interopDefault` option to `true` with `jiti` so that it is possible for them to do that. They can even use the `import` syntax with `module.exports`. But the reason I presented the question is that to my knowledge that shouldn't be a problem there, but CJS/ESM compatibility can be a can of worms and I wanted to make sure that it's something we might need to keep in mind.

## Help Needed

<!--
    This section is optional.

    Are you able to implement this RFC on your own? If not, what kind
    of help would you need from the team?
-->

I will be implementing this feature. I will need help from the team to review the code and provide feedback.

## Frequently Asked Questions

<!--
    This section is optional but suggested.

    Try to anticipate points of clarification that might be needed by
    the people reviewing this RFC. Include those questions and answers
    in this section.
-->

> ## Related Discussions

Reviewer: More prior discussions worth linking to, as @aladdin-add mentioned:

* https://github.com/eslint/eslint/issues/12078
* https://github.com/eslint/rfcs/pull/50

Author: I linked them, thanks!

<!-- One-paragraph explanation of the feature. -->

## Motivation

<!-- Why are we doing this? What use cases does it support? What is the expected
outcome? -->

The primary motivation for adding support for TypeScript configuration files to ESLint is to enhance the developer experience and accommodate the evolving JavaScript ecosystem. As TypeScript's popularity continues to grow, more projects are adopting TypeScript not only for their source code but also for their configuration files. This shift is driven by TypeScript's ability to provide compile-time type checks and IntelliSense. By supporting `eslint.config.ts`, `eslint.config.mts`, and `eslint.config.cts`, ESLint will offer first-class support to TypeScript users, allowing them to leverage these benefits directly within their ESLint configuration.

## Detailed Design

<!--
   This is the bulk of the RFC.

   Explain the design with enough detail that someone familiar with ESLint
   can implement it by reading this document. Please get into specifics
   of your approach, corner cases, and examples of how the change will be
   used. Be sure to define any new terms in this section.
-->

> The goal is to seamlessly support TypeScript configuration files in ESLint. To achieve this, ESLint will need to recognize and parse TypeScript configuration files in the same way it does for JavaScript configuration files. This will involve creating the configuration file resolution logic to recognize `.ts`, `.mts`, and `.cts` files as valid ESLint configuration files. We will need to treat `eslint.config.mts` files as ECMAScript Modules (ESM) and `eslint.config.cts` files as CommonJS (CJS). The module resolution for `eslint.config.ts` files will depend on the `type` field in the nearest `package.json`. The maintainers of ESLint have raised some valid concerns some of which include:

Reviewer 3: Can you give an example of how eslint configs should be typed?  as of now, eslint does not export typings.

tools like vite support an api: `defineConfig()`, should we do the similar thing?

```ts
import "defineConfig" from "eslint";

export default defineConfig([
....
]);
```

Author: I would absolutely love that, but that's up to the maintainers, I would be happy to include a `defineConfig()` or `createESLintConfig()` in the linked PR or even in a separate PR. All I need is the thumbs up!

Author: Just for reference, I have something like that setup in [Redux-Toolkit](https://github.com/reduxjs/redux-toolkit/blob/94fb3a77f95f8f10e939c32d7a2050511a786bdd/packages/toolkit/eslint.config.mjs).

In an ESM file it can look like this:

```ts
import { createESLintConfig } from '@reduxjs/eslint-config'

export default createESLintConfig([
  {
    rules: {
      'no-console': [0],
    },
  },
  {
    // ...Other additional overrides
  },
])
```

and in a CJS file it can look like this:

```ts
module.exports = (async () =>
  (await import('@reduxjs/eslint-config')).createESLintConfig([
    {
      rules: {
        'no-console': [0],
      },
    },
    {
      // ...Other additional overrides
    },
  ]))()
```

Reviewer 3: yes, it's worthy to be listed in the open questions.

Author: Done!

- There should not be extra overhead for JavaScript users. This means this change should not have a significant impact (if any at all) affecting users who use plain JavaScript config files.
- The external tools that are used to parse the config files written in TypeScript should not create side effects. Specifically, it is imperative that these tools do not interfere with Node.js's native module resolution system by hooking into or altering the standard `import/require` mechanisms. This means tools like [`ts-node`](https://github.com/TypeStrong/ts-node) and [`tsx`](https://github.com/privatenumber/tsx) might not be suitable for this purpose.

So far the tool that seems to be the most suitable for this purpose is [`jiti`](https://www.npmjs.com/package/jiti). It does not introduce side effects and performs well, demonstrating its reliability. It also seems to be more battle-tested given some established frameworks such as [Nuxt](https://github.com/nuxt/nuxt) and [Tailwind CSS](https://github.com/tailwindlabs/tailwindcss) have been using it to load their configuration files.

> ## Examples

Reviewer: I think it would be good context to provide to call out that you can perform type-checking already with the [`checkJs`](https://www.typescriptlang.org/tsconfig#checkJs) TypeScript option and a JS config file:

```js
/** @type {import('eslint').Linter.FlatConfig[]} */
const config = [/* ... */];

module.exports = config;
```

An actual TypeScript config file has other benefits besides just type-checking, but calling out alternatives/workarounds to achieve the desired behavior/functionality can be useful in weighing the benefits of the proposal.

Author: Sure thing!

3. Should we allow some sort of interoperability between JavaScript and TypeScript configuration files? For example, should we allow a TypeScript configuration file to extend a JavaScript configuration file and vice versa?
4. Should we allow `eslint.config.ts` to be able to use `export default` as well as `module.exports` (might be related to [TypeScript's automatic Module Detection](https://www.typescriptlang.org/tsconfig#moduleDetection))?

## Help Needed

<!--
    This section is optional.

    Are you able to implement this RFC on your own? If not, what kind
    of help would you need from the team?
-->

I will be implementing this feature. I will need help from the team to review the code and provide feedback.

## Frequently Asked Questions

<!--
    This section is optional but suggested.

    Try to anticipate points of clarification that might be needed by
    the people reviewing this RFC. Include those questions and answers
    in this section.
-->

## Related Discussions

<!--
    This section is optional but suggested.

    If there is an issue, pull request, or other URL that provides useful
    context for this proposal, please include those links here.
-->

> [This PR](https://github.com/eslint/eslint/pull/18134) is related to this RFC.
> [Prior Discussion](https://github.com/eslint/rfcs/pull/50) related to supporting `.eslintrc.ts` files.
> [Prior Issue](https://github.com/eslint/eslint/issues/12078) related to supporting `.eslintrc.ts` files.

Reviewer: Better readability as a list:

```suggestion
* [This PR](https://github.com/eslint/eslint/pull/18134) is related to this RFC.
* [Prior Discussion](https://github.com/eslint/rfcs/pull/50) related to supporting `.eslintrc.ts` files.
* [Prior Issue](https://github.com/eslint/eslint/issues/12078) related to supporting `.eslintrc.ts` files.
```

Author: Done! Thank you!

- There should not be extra overhead for JavaScript users. This means this change should not have a significant impact (if any at all) affecting users who use plain JavaScript config files.
- The external tools that are used to parse the config files written in TypeScript should not create side effects. Specifically, it is imperative that these tools do not interfere with Node.js's native module resolution system by hooking into or altering the standard `import/require` mechanisms. This means tools like [`ts-node`](https://github.com/TypeStrong/ts-node) and [`tsx`](https://github.com/privatenumber/tsx) might not be suitable for this purpose.

> So far the tool that seems to be the most suitable for this purpose is [`jiti`](https://www.npmjs.com/package/jiti). It does not introduce side effects and performs well, demonstrating its reliability. It also seems to be more battle-tested given some established frameworks such as [Nuxt](https://github.com/nuxt/nuxt), [Tailwind CSS](https://github.com/tailwindlabs/tailwindcss) and [Docusaurus](https://github.com/facebook/docusaurus) have been using it to load their configuration files.

Reviewer 2: As part of the detailed design, please explain how and where you'd use Jiti in the ESLint codebase.

Author: Done!

2. Using [`tsx`](https://github.com/privatenumber/tsx) to parse TypeScript configuration files. This approach also proved to be problematic because [`tsx`](https://github.com/privatenumber/tsx) hooks into Node.js's native module resolution system, which could potentially cause side effects.

3. Using [TypeScript's `transpileModule()`](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API#a-simple-transform-function) to parse TypeScript configuration files. This approach proved to be problematic because it requires a significant amount of overhead and is not suitable for this purpose.

## Open Questions

<!--
    This section is optional, but is suggested for a first draft.

    What parts of this proposal are you unclear about? What do you
    need to know before you can finalize this RFC?

    List the questions that you'd like reviewers to focus on. When
    you've received the answers and updated the design to reflect them,
    you can remove this section.
-->

> 1. How is caching going to work with TypeScript config files?

Reviewer 2: We only cache the computed result of loading a config file, so I don't think this should be a problem.

> 2. Should we look at the nearest `tsconfig.json` file to determine the module resolution for `eslint.config.ts` files? Most likely not, but it's worth considering.

Reviewer 2: Are you asking if ESLint needs to locate and load the `tsconfig.json` file? If so, I'd definitely prefer not to do that. I'd like ESLint to be as TypeScript-ignorant as possible.

Author: I agree!

> 3. Should we allow some sort of interoperability between JavaScript and TypeScript configuration files? For example, should we allow a TypeScript configuration file to extend a JavaScript configuration file and vice versa?

Reviewer 2: I don't believe this is an issue as the `extends` key isn't supported. People just use `import` to load anything else they need.

4. Should we allow `eslint.config.ts` to be able to use `export default` as well as `module.exports` (might be related to [TypeScript's automatic Module Detection](https://www.typescriptlang.org/tsconfig#moduleDetection))?

> 5. Tools like [Vitest](https://github.com/vitest-dev/vitest) export a [`defineConfig`](https://vitest.dev/config/file.html#managing-vitest-config-file) function to make it easier to write configuration files in TypeScript. Should we consider doing something similar for ESLint?

Reviewer 2: We could consider that separately, but I don't see it as important for this RFC.

<!-- One-paragraph explanation of the feature. -->

## Motivation

<!-- Why are we doing this? What use cases does it support? What is the expected
outcome? -->

The primary motivation for adding support for TypeScript configuration files to ESLint is to enhance the developer experience and accommodate the evolving JavaScript ecosystem. As TypeScript's popularity continues to grow, more projects are adopting TypeScript not only for their source code but also for their configuration files. This shift is driven by TypeScript's ability to provide compile-time type checks and IntelliSense. By supporting `eslint.config.ts`, `eslint.config.mts`, and `eslint.config.cts`, ESLint will offer first-class support to TypeScript users, allowing them to leverage these benefits directly within their ESLint configuration.

## Detailed Design

<!--
   This is the bulk of the RFC.

   Explain the design with enough detail that someone familiar with ESLint
   can implement it by reading this document. Please get into specifics
   of your approach, corner cases, and examples of how the change will be
   used. Be sure to define any new terms in this section.
-->

> The goal is to seamlessly support TypeScript configuration files in ESLint. To achieve this, ESLint will need to recognize and parse TypeScript configuration files in the same way it does for JavaScript configuration files. This will involve creating the configuration file resolution logic to recognize `.ts`, `.mts`, and `.cts` files as valid ESLint configuration files. We will need to treat `eslint.config.mts` files as ECMAScript Modules (ESM) and `eslint.config.cts` files as CommonJS (CJS). The module resolution for `eslint.config.ts` files will depend on the `type` field in the nearest `package.json`. The maintainers of ESLint have raised some valid concerns some of which include:

Reviewer 4: Again, it's possible that I'm overlooking something but it seems that Jiti makes no difference between `.ts`, `.mts` and `.cts` files, as these are all transpiled to CommonJS, for example:

```ts
export default [];

console.log(typeof module);      // prints "object"
console.log(typeof require);     // prints "function"
console.log(typeof exports);     // prints "object"
console.log(typeof __filename);  // prints "string"
console.log(typeof __dirname);   // prints "string"
```

The output is consistent with CommonJS, and it doesn't matter if this file is named `eslint.config.ts`, `eslint.config.mts` or `eslint.config.cts`.

If my analysis is not correct, could you add a reproducible example to show the differences between the file formats?

Author: > Again, it's possible that I'm overlooking something but it seems that Jiti makes no difference between `.ts`, `.mts` and `.cts` files, as these are all transpiled to CommonJS, for example:
> 
> ```ts
> export default [];
> 
> console.log(typeof module);      // prints "object"
> console.log(typeof require);     // prints "function"
> console.log(typeof exports);     // prints "object"
> console.log(typeof __filename);  // prints "string"
> console.log(typeof __dirname);   // prints "string"
> ```
> 
> The output is consistent with CommonJS, and it doesn't matter if this file is named `eslint.config.ts`, `eslint.config.mts` or `eslint.config.cts`.
> 
> If my analysis is not correct, could you add a reproducible example to show the differences between the file formats?

You are correct, `jiti` handles everything synchronously hence why it doesn't support top level `await`. It transforms everything to CommonJS which means users can use either `module.exports` or `export default` and that level of interoprability, is exactly why `jiti` is so widely used _specifically_ for handling config files.

> The goal is to seamlessly support TypeScript configuration files in ESLint. To achieve this, ESLint will need to recognize and parse TypeScript configuration files in the same way it does for JavaScript configuration files. This will involve creating the configuration file resolution logic to recognize `.ts`, `.mts`, and `.cts` files as valid ESLint configuration files. We will need to treat `eslint.config.mts` files as ECMAScript Modules (ESM) and `eslint.config.cts` files as CommonJS (CJS). The module resolution for `eslint.config.ts` files will depend on the `type` field in the nearest `package.json`. The maintainers of ESLint have raised some valid concerns some of which include:

Reviewer 4: ```suggestion
The goal is to seamlessly support TypeScript configuration files in ESLint. To achieve this, ESLint will need to recognize and parse TypeScript configuration files in the same way it does for JavaScript configuration files. This will involve creating the configuration file resolution logic to recognize `.ts`, `.mts`, and `.cts` files as valid ESLint configuration files. We will need to treat these files as TypeScript files. The resolution of imported modules will depend on the `type` field in the nearest `package.json`, and users will be able to specify exports using either `module.exports` or the `export default` syntax regardless of the config file extension. The maintainers of ESLint have raised some valid concerns some of which include:
```

Maybe like this, to make it clear that the file extensions is irrelevant to the type of module being generated. Also, "module resolution" could be easily confused with the TypeScript [`moduleResolution`](https://www.typescriptlang.org/tsconfig#moduleResolution) options, which isn't meat here, so it's better to use a different wording.

Author: Done!

## Open Questions

<!--
    This section is optional, but is suggested for a first draft.

    What parts of this proposal are you unclear about? What do you
    need to know before you can finalize this RFC?

    List the questions that you'd like reviewers to focus on. When
    you've received the answers and updated the design to reflect them,
    you can remove this section.
-->

1. How is caching going to work with TypeScript config files? We only cache the computed result of loading a config file, so I don't think this should be a problem.
2. Should we look at the nearest `tsconfig.json` file to determine the module resolution for `eslint.config.ts` files? Most likely not, but it's worth considering.
3. Should we allow some sort of interoperability between JavaScript and TypeScript configuration files? For example, should we allow a TypeScript configuration file to extend a JavaScript configuration file and vice versa? I don't believe this is an issue as the `extends` key isn't supported. Users just use `import` to load anything else they need.

> 4. Should we allow `eslint.config.ts` to be able to use `export default` as well as `module.exports` (might be related to [TypeScript's automatic Module Detection](https://www.typescriptlang.org/tsconfig#moduleDetection))?

Reviewer 4: I'd say it's okay to allow both. TypeScript will already complain if the export style doesn't match the expected module type, and Jiti doesn't make a difference between ESM and CommonJS anyway.

Author: I agree!

<!-- One-paragraph explanation of the feature. -->

## Motivation

<!-- Why are we doing this? What use cases does it support? What is the expected
outcome? -->

The primary motivation for adding support for TypeScript configuration files to ESLint is to enhance the developer experience and accommodate the evolving JavaScript ecosystem. As TypeScript's popularity continues to grow, more projects are adopting TypeScript not only for their source code but also for their configuration files. This shift is driven by TypeScript's ability to provide compile-time type checks and IntelliSense. By supporting `eslint.config.ts`, `eslint.config.mts`, and `eslint.config.cts`, ESLint will offer first-class support to TypeScript users, allowing them to leverage these benefits directly within their ESLint configuration.

## Detailed Design

<!--
   This is the bulk of the RFC.

   Explain the design with enough detail that someone familiar with ESLint
   can implement it by reading this document. Please get into specifics
   of your approach, corner cases, and examples of how the change will be
   used. Be sure to define any new terms in this section.
-->

> The goal is to seamlessly support TypeScript configuration files in ESLint. To achieve this, ESLint will need to recognize and parse TypeScript configuration files in the same way it does for JavaScript configuration files. This will involve creating the configuration file resolution logic to recognize `.ts`, `.mts`, and `.cts` files as valid ESLint configuration files. We will need to treat these files as TypeScript files. The resolution of imported modules will depend on the `type` field in the nearest `package.json`, and users will be able to specify exports using either `module.exports` or the `export default` syntax regardless of the config file extension. The maintainers of ESLint have raised some valid concerns some of which include:

Author: It does but only when dealing with a plain `".js"` extension so in our case no. I'll fix it.

- There should not be extra overhead for JavaScript users. This means this change should not have a significant impact (if any at all) affecting users who use plain JavaScript config files.
- The external tools that are used to parse the config files written in TypeScript should not create side effects. Specifically, it is imperative that these tools do not interfere with Node.js's native module resolution system by hooking into or altering the standard `import/require` mechanisms. This means tools like [`ts-node`](https://github.com/TypeStrong/ts-node) and [`tsx`](https://github.com/privatenumber/tsx) might not be suitable for this purpose.

> So far the tool that seems to be the most suitable for this purpose is [`jiti`](https://www.npmjs.com/package/jiti). It does not introduce side effects and performs well, demonstrating its reliability. It also seems to be more battle-tested given some established frameworks such as [Nuxt](https://github.com/nuxt/nuxt), [Tailwind CSS](https://github.com/tailwindlabs/tailwindcss) and [Docusaurus](https://github.com/facebook/docusaurus) have been using it to load their configuration files.

Reviewer 5: I would like to propose using `tsx` instead of `jiti`. Implementation PR is here: https://github.com/eslint/eslint/pull/18440

While being a Nuxt team member and having `jiti` users, I'd be really happy to see `jiti` being used. However, I'd say that `jiti`'s approach is not honestly future-proof. It uses a built-in babel parser to transpile TS and ESM code and evaluate them in CJS mode, which means it does not support top-level await, and could have some misalignment with ESM. While there is a plan to support full ESM mode, the implementation isn't easy and I would not expect it to be landed very soon (I tried to make it happend a few times but didn't work out)

On the other hand, [`tsx`'s recent `tsx/esm/api`](https://tsx.is/node#tsimport) seems like a much better approach as it uses Node's native loader API, which means the module resolutions and evaluation is in native Node, which would have a much more smaller interface of potential misalignment.

I expiremented three different loader approaches in [`eslint-ts-patch`](https://github.com/antfu/eslint-ts-patch), and so far I see the `tsx` the most solid approach. The only downside is that it requires minimal Node.js v20.8.0 and v18.19.0, but I suppose it won't be an issue very soon as the ecosystem moving forward.

   * if it has been changed since the last time it was imported.
   * If it hasn't been changed, `import()` will just return the cached version.
   *
   * Note that we should not overuse queries (e.g., by appending the current time
   * to always reload the config file module) as that could cause memory leaks
   * because entries are never removed from the import cache.
   */
  fileURL.searchParams.append("mtime", mtime)

  /*
   * With queries, we can bypass the import cache. However, when import-ing a CJS module,
   * Node.js uses the require infrastructure under the hood. That includes the require cache,
   * which caches the config file module by its file path (queries have no effect).
   * Therefore, we also need to clear the require cache before importing the config file module.
   * In order to get the same behavior with ESM and CJS config files, in particular - to reload
   * the config file only if it has been changed, we track file modification times and clear
   * the require cache only if the file has been changed.
   */
  if (importedConfigFileModificationTime.get(filePath) !== mtime) {
    delete require.cache[filePath]
  }

  const isTS = isFileTS(filePath)

>   if (isTS) {

Reviewer 2: We should probably do a simple test to see if Jiti is needed (it's not in Deno or Bun):

```suggestion
  if (isTS && !globalThis.Deno && !globalThis.Bun) {
```

    Any change has potential downsides, including increased maintenance
    burden, incompatibility with other tools, breaking existing user
    experience, etc. Try to identify as many potential problems with
    implementing this RFC as possible.
-->

This change will most likely require at least one external tool as either a dependency or a peer dependency.

## Backwards Compatibility Analysis

<!--
    How does this change affect existing ESLint users? Will any behavior
    change for them? If so, how are you going to minimize the disruption
    to existing users?
-->

This goal is to minimize the disruption to existing users. The primary focus is to ensure that the existing JavaScript configuration files continue to work as expected. The changes will only affect TypeScript users who are using TypeScript configuration files. The changes will not affect the existing JavaScript configuration files.

## Alternatives

<!--
    What other designs did you consider? Why did you decide against those?

    This section should also include prior art, such as whether similar
    projects have already implemented a similar feature.
-->

> 

Reviewer 3: there is a community project: https://github.com/antfu/eslint-ts-patch to support it.  would like to hear the author :) / @antfu

Reviewer 5: It uses [jiti](https://github.com/unjs/jiti) to support ts file. Which already mentioned https://github.com/eslint/rfcs/pull/117#issuecomment-2036336411 that it doesn't currently support top-level await.

I would personally recommend using https://github.com/egoist/bundle-require instead which is more robust and will respect tsconfig.json. The downside is that it would introduce `esbuild` into the dependency. If the install size is a concern, I guess we could have an optional package like `@eslint/config-loader-ts` that only requires when users use ts version of config.

Reviewer 5: Yes, the inspector uses that because we need to know the dependencies of the config to do automatic reloads. Supporting TS was a free side-effect. 

Even if ESLint doesn't need information on dependencies, I think it's still a good way to support TS. Vite uses the same approach to load `vite.config.ts`.

Reviewer 6: `bundle-require` writes a temp file besides the original right? These temp files are a major source of problems with vite, see https://github.com/vitejs/vite/issues/9470.

Reviewer 2: @antfu 

> Does bundle-require write a temp file to disk like vite does?

Reviewer 5: Yes it does. But I think currently there is no better solution unless Node supports TS natively, or we pass the loader flag in our CLI: https://tsx.is/node#es-modules-only

Reviewer 2: So it sounds like the deciding factor here is whether or not we think that top-level await is important to support in TypeScript config files. If it *is* important, then `bundle-require` is the only option with the caveat that it creates a temp file; if it *isn't* important, then we can use Jiti.

What are people's opinions on whether or not top-level await is important?

Reviewer 6: Can't think of any case where I wold use TLA in any config file. If I need to do file I/O, I would do is synchronous because that is faster than the async variants. If a user would want to `fetch` it would likely be unavoidable to have some support for async.

You could also avoid TLA by accepting a function that return either a promise or sync result, e.g.

```js
export default async () => {files: []};
export default () => {files: []};

Author: I agree with @silverwind here, I think [`jiti`](https://www.npmjs.com/package/jiti) would be more appropriate in this case since top-level `await` expressions are not something I see people using in an ESLint config file. [`jiti`](https://www.npmjs.com/package/jiti) also has the benfit of ignoring `tsconfig.json` files which according to [this comment](https://github.com/eslint/rfcs/pull/117#discussion_r1543602282) is the desired outcome here. It also doesn't write temp files to disk which as mentioned in [this comment](https://github.com/eslint/rfcs/pull/117#discussion_r1561318636) can potentially cause some issues.

Reviewer 7: tsx v4.9 exports `tsImport()` which supports loading TypeScript files with TLA (compiles in-memory).

https://tsx.is/node#enhanced-import-require

Author: @privatenumber That's wonderful news I'll give it a shot.

Reviewer 2: I agree that TLA shouldn't be a blocker for this. :+1:

Reviewer 4: tsx is mentioned as a non-viable alternative because it uses custom loaders. Does that also hold when `tsImport()` is used?

Reviewer 7: Personally, I see TLA as a language spec requirement, and I don't think whether to support it should be up to whether we can imagine use-cases.

But if we want evaluate usage, here are examples of TLA in configs (Vite, Rollup, Next, Jest, esbuild, etc):
https://github.com/search?q=path%3A%2F%5C.config%5C.%28js%7Cts%29%24%2F+%2F%5Eawait+%2F&type=code

----

> tsx is mentioned as a non-viable alternative because it uses custom loaders. Does that also hold when `tsImport()` is used?

`tsImport()` is isolated and won't add TypeScript support to the rest of the environment.

But just to be clear, I'm only sharing it as an option. Personally, I would like to have popular compilers supported as optional peer dependencies to accommodate different preferences and environments.

Reviewer 4: > Personally, I see TLA as a language spec requirement, and I don't think whether to support it should be up to whether we can imagine use-cases.
> 
> But if we want evaluate usage, here are examples of TLA in configs (Vite, Rollup, Next, Jest, esbuild, etc): https://github.com/search?q=path%3A%2F%5C.config%5C.%28js%7Cts%29%24%2F+%2F%5Eawait+%2F&type=code

Interesting. There are also ESLint configs that use TLA: https://github.com/search?q=path%3A%2Feslint%5C.config%5C.js%24%2F+%2F%5Econst+.%2B+%3D+await+%2F&type=code.

Reviewer 4: Jiti won't recognize TLA not only in the config file itself, but also in imported modules, because they are all treated as CommonJS. I did a quick test using a patched version of ESLint with the changes from [#18134](https://github.com/eslint/eslint/pull/18134) and with this config:

```ts
// eslint.config.ts
export { default } from './recommended.mjs';
```

```js
// recommended.mjs
import { readFile } from 'node:fs/promises';

const json = await readFile('package.json', 'utf-8');
const { name } = JSON.parse(json);

export default [{ name, rules: { 'no-undef': 'error' } }];
```

When running `eslint`, I got an error as expected:

```
Oops! Something went wrong! :(

ESLint: 9.2.0

ReferenceError: await is not defined
    at ../project/recommended.mjs:4:14
    at evalModule (../project/node_modules/eslint/node_modules/jiti/dist/jiti.js:1:256443)
    at jiti (../project/node_modules/eslint/node_modules/jiti/dist/jiti.js:1:254371)
    at ../project/eslint.config.ts:2:43
    at evalModule (../project/node_modules/eslint/node_modules/jiti/dist/jiti.js:1:256443)
    at jiti (../project/node_modules/eslint/node_modules/jiti/dist/jiti.js:1:254371)
    at loadFlatConfigFile (../project/node_modules/eslint/lib/eslint/eslint.js:335:24)
    at async calculateConfigArray (../project/node_modules/eslint/lib/eslint/eslint.js:421:28)
    at async ESLint.lintFiles (../project/node_modules/eslint/lib/eslint/eslint.js:840:25)
    at async Object.execute (../project/node_modules/eslint/lib/cli.js:500:23)
    at async main (../project/node_modules/eslint/bin/eslint.js:153:22)
```

The workaround of using a dynamic import didn't help either, and resulted in the same error:

```ts
// eslint.config.ts
export default (async () =>
    (await import('./recommended.mjs')).default
)();
```

This means that if we decide to use Jiti, plugin developers should be advised to avoid TLA in their shared configs (including rules and transitive dependencies), or else those configs will not work for users who have a `eslint.config.ts` in their project.

Reviewer 5: I added support for all the ts loaders we mentioned in [`eslint-ts-patch`](https://github.com/antfu/eslint-ts-patch#typescript-loaders) and listed their trade-offs, where you can give them a try today.

Reviewer 4: @privatenumber I tried to use tsx `tsImport` as explained [here](https://github.com/privatenumber/tsx/blob/master/docs/node.md) to load a `.ts` config file, but didn't succeed. This is my code:

```js
let config;
const { tsImport, register } = require("tsx/esm/api");
const unregister = register();
try {
    config = (await tsImport(fileURL.href, __filename)).default;
} finally {
    unregister();
}
```

where `fileURL` is the URL of this config file:

```ts
// eslint.config.mts
export default [];
```

The error stack trace indicates that the config file is being loaded as CommonJS:

```
SyntaxError: Unexpected token 'export'
    at wrapSafe (node:internal/modules/cjs/loader:1389:18)
    at Module._compile (node:internal/modules/cjs/loader:1425:20)
    at Module._extensions..js (node:internal/modules/cjs/loader:1564:10)
    at Module.load (node:internal/modules/cjs/loader:1287:32)
    at Module._load (node:internal/modules/cjs/loader:1103:12)
    at cjsLoader (node:internal/modules/esm/translators:318:15)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:258:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:262:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:474:24)
```

I guess I'm doing something wrong?

Reviewer 7: This is probably not the best place to discuss this. Can you send me a reproduction link in the tsx repo?

Reviewer 4: > This is probably not the best place to discuss this. Can you send me a reproduction link in the tsx repo?

Thanks! I think the problem is that I wasn't calling `register` from `require("tsx/cjs/api")`. It works if I add that call before `tsImport`. Another problem is that there is seemingly no way to completely undo the changes done by `tsImport` to the Node.js loader internals. Particularly, I think there is no way to unregister the [ESM `Module` loader](https://github.com/privatenumber/tsx/blob/4f515ab0bb8f3c8c610dc57556b978cbf90a5560/src/esm/api/register.ts#L22-L33). I didn't even find out how to do that programmatically, so this is possibly a limitation of Node.js.

     * because entries are never removed from the import cache.
     */
    fileURL.searchParams.append("mtime", mtime);

    /*
     * With queries, we can bypass the import cache. However, when import-ing a CJS module,
     * Node.js uses the require infrastructure under the hood. That includes the require cache,
     * which caches the config file module by its file path (queries have no effect).
     * Therefore, we also need to clear the require cache before importing the config file module.
     * In order to get the same behavior with ESM and CJS config files, in particular - to reload
     * the config file only if it has been changed, we track file modification times and clear
     * the require cache only if the file has been changed.
     */
    if (importedConfigFileModificationTime.get(filePath) !== mtime) {
        delete require.cache[filePath];
    }

    const isTS = isFileTS(filePath);

    if (isTS) {
        const jiti = (await import("jiti")).default(__filename, { interopDefault: true });

        const config = jiti(fileURL.href);

        return config;
    }

    const config = (await import(fileURL.href)).default;

>     importedConfigFileModificationTime.set(filePath, mtime);

Author: I fixed it, thanks for pointing it out.
